protocols/upnp/src/behaviour.rs

```diff
@@ -0,0 +1,554 @@
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        _connection_id: ConnectionId,
+        _peer: PeerId,
+        _addr: &Multiaddr,
+        _role_override: Endpoint,
+    ) -> Result<libp2p_swarm::THandler<Self>, libp2p_swarm::ConnectionDenied> {
+        Ok(dummy::ConnectionHandler)
+    }
+
+    fn on_swarm_event(&mut self, event: FromSwarm<Self::ConnectionHandler>) {
+        match event {
+            FromSwarm::NewListenAddr(NewListenAddr {
+                listener_id,
+                addr: multiaddr,
+            }) => {
+                let (addr, protocol) = match multiaddr_to_socketaddr_protocol(multiaddr.clone()) {
+                    Ok(addr_port) => addr_port,
+                    Err(_) => {
+                        log::debug!("multiaddress not supported for UPnP {multiaddr}");
+                        return;
+                    }
+                };
+
+                if let Some((mapping, _state)) = self
+                    .mappings
+                    .iter()
+                    .find(|(mapping, _state)| mapping.internal_addr.port() == addr.port())
+                {
+                    log::debug!("port from multiaddress {multiaddr} is already being mapped to another multiaddr: {}", mapping.multiaddr);
+                    return;
+                }
+
+                match &mut self.state {
+                    GatewayState::Searching(_) => {
+                        // As the gateway is not yet available we add the mapping with `MappingState::Inactive`
+                        // so that when and if it becomes available we map it.
+                        self.mappings.insert(
+                            Mapping {
+                                listener_id,
+                                protocol,
+                                internal_addr: addr,
+                                multiaddr: multiaddr.clone(),
+                            },
+                            MappingState::Inactive,
+                        );
+                    }
+                    GatewayState::Available(ref mut gateway) => {
+                        let mapping = Mapping {
+                            listener_id,
+                            protocol,
+                            internal_addr: addr,
+                            multiaddr: multiaddr.clone(),
+                        };
+
+                        let duration = self.config.temporary.then_some(MAPPING_DURATION);
+                        if let Err(err) = gateway
+                            .sender
+                            .try_send(GatewayRequest::AddMapping(mapping.clone(), duration))
+                        {
+                            log::debug!(
+                                "could not request port mapping for {} on the gateway: {}",
+                                mapping.multiaddr,
+                                err
+                            );
+                        }
```

Comment:
Note to myself: think about a way to do without the `try_send`. Definitely better than an unbounded channel or an unbounded request buffer (e.g. `Vec<GatewayRequest>`). Need to think about the retry behavior on drop.

Reply:
With the current architecture, I don't think we need a way around the `try_send`. In other words, in order to not block the main event loop (`NetworkBehaviour`) and in order to not accumulate unbounded memory, we have to drop somewhere. This seems to me the right place. With that in mind I am resolving this conversation.